### PR TITLE
update: Qianfan dataset optimization with a few new functions

### PR DESCRIPTION
**Pull request by Dobiichi-Origami**
This PR comes up with an improved `dataset.Table` implementation specific to nested data, and introduces new apis in `resources.console.data` including ETL and Augmentation for dataset

Improvement:
+ optimize the representation of nested array-like dataset in dataset, make performance better and operation more user-friendly
+ schema has been rewrite for more accurate validation, without complex logic
+ along with the new representation of dataset, now the code of load/save have been more robust

Feature:
+ new console api about dataset etl, augmentation and annotation
+ support releasing on Qianfan data source, even if after uploading
+ create an etl task from sdk using `online_data_process ` with `QianfanOperator`
+ check your dataset data without downloading ( a.k.a preview ) by `list`

### CODE REVIEW

File: src/qianfan/dataset/dataset.py [hunk 1]
```diff
@@ -157,15 +156,17 @@ def _to_source(self, source: DataSource, **kwargs: Any) -> bool:
...
-                    QianfanDefaultColumnNameForNestedTable
-                ]
+            if QianfanDataGroupColumnName in column_names:
+                compo_list: List[List[Dict[str, Any]]] = []
+                for row in self.inner_table.to_pylist():
+                    group_index = row[QianfanDataGroupColumnName]
+                    while group_index >= len(compo_list):
+                        compo_list.append([])
+                    row.pop(QianfanDataGroupColumnName)
+                    compo_list[-1].append(row)
```
Comment (stonekim) [inline src/qianfan/dataset/dataset.py:168]: -1 还是 group_index?

File: src/qianfan/dataset/dataset.py [hunk 2]
```diff
@@ -434,11 +435,67 @@ def _is_dataset_located_in_qianfan(self) -> bool:
...
+
+        log_info(f"created etl task id: {etl_id}")
+        return etl_id
+
     # 因为要针对数据集在千帆上的在线处理，所以需要加一些额外的处理逻辑。
     # 例如通过平台的 API 发起数据清洗任务
     def online_data_process(self, operators: List[QianfanOperator]) -> bool:
         """
-        create a online ETL task on qianfan
+        create an online ETL task on qianfan
```
Comment (stonekim) [inline src/qianfan/dataset/dataset.py:498]: 这个函数返回值可以增加多些信息：etl_id和对应处理后的datasetid组成

File: src/qianfan/dataset/dataset.py [hunk 3]
```diff
@@ -513,12 +609,10 @@ def append(self, elem: Any) -> Self:
         return super().append(elem)
 
     # 等待接口 ready 才能对云端数据集做展示
-    @_online_except_decorator
     def list(
         self,
-        by: Optional[
-            Union[slice, int, str, List[int], Tuple[int], List[str], Tuple[str]]
-        ] = None,
+        by: Optional[Union[slice, int, str, Sequence[int], Sequence[str]]] = None,
```
Comment (stonekim) [inline src/qianfan/dataset/dataset.py:614]: 函数上面的注释

File: src/qianfan/dataset/dataset.py (hunk 1 above)
Comment (Dobiichi-Origami) [reply src/qianfan/dataset/dataset.py:168]: -1 和 group_index 都可以，因为当前处理的 group 永远是 compo_list 的最后一个列表元素

File: src/qianfan/dataset/dataset.py (hunk 3 above)
Comment (Dobiichi-Origami) [reply src/qianfan/dataset/dataset.py:614]: 由于 mypy 要求对齐函数签名，因此这里得写成接口的函数签名
但是这个函数实际上不能够接受所有这些参数
实际参数还是得以签名为准，这样还需要改吗

**Review by Dobiichi-Origami — COMMENTED** [review]
report

File: src/qianfan/dataset/dataset.py [hunk 4]
```diff
@@ -434,27 +435,147 @@ def _is_dataset_located_in_qianfan(self) -> bool:
...
+            Dict[str, Any]: ETL task info, contains 3 field:
+                is_succeeded (bool): whether ETL task succeed
+                etl_task_id (Optional[int]): etl task id, only
+                    exists when etl task is created successfully
+                new_dataset_id (Optional[int]): dataset id which
+                    stores data after etl, only exists when etl
+                    task is succeeded
         """
+
+        ret_dict: Dict[str, Any] = {"is_succeeded": False}
```
Comment (stonekim) [inline src/qianfan/dataset/dataset.py:517]: 没看到设为 True 的过程

Comment (Dobiichi-Origami) [reply src/qianfan/dataset/dataset.py:517]: > 没看到设为 True 的过程

here
截屏2023-11-23 20 51 33

**Review by stonekim — APPROVED** [review]

**Pull request merged**